skorch/utils.py

```diff
@@ -632,7 +633,7 @@ def _infer_predict_nonlinearity(net):
     if isinstance(criterion, CrossEntropyLoss):
         return partial(torch.softmax, dim=-1)
 
-    if isinstance(criterion, BCEWithLogitsLoss):
+    if isinstance(criterion, (BCEWithLogitsLoss, BCELoss)):
```

Comment:
In the test, `module__output_nonlin=torch.nn.Sigmoid()`, which means the output already has sigmoid applied. I think using this non-linearity would apply another sigmoid.

If we assume that the module already does sigmoid then the module already outputs probabilities. In that case, I think the non-linearity should only be `torch.stack((1 - prob, prob), 1)` part of `_sigmoid_then_2d`.

Reply:
Very good catch, thanks. I fixed the bug by not applying the sigmoid in the postprocessing step in `utils.py`. We still need it as output nonlinearity because that is what `BCELoss` expects.

To catch this kind of bug in the future, I extended the unit tests. When sigmoid is applied twice, all probabilities would be in the range of [sigmoid(-1), sigmoid(1)]. I now check that at least some returned probabilities are outside that range.